### ISSUE TITLE
Product Bundles total cost in Google Analytics doesn't reflect bundled items 

### ISSUE THREAD

**Issue by selectedselections** [woocommerce/woocommerce-google-analytics-integration#219]
### Describe the bug:

When a product bundle is passed to Google, the base cost is considered, but not the bundled items. If the "base cost" is set to `0` and the items within the bundle are set to `priced individually` then the result is that the price will be shown as `0` in Google, when it should be the total of items in the bundle. 

### Steps to reproduce:

1. Create a product bundle using [Product Bundles](https://woocommerce.com/products/product-bundles/?aff=10486&cid=1131038) 
2. Set the cost of the product to 0 [screenshot](https://d.pr/i/5pRT7K)
3. Add bundled items and make sure theyre set to "priced individually" [screenshot](https://d.pr/i/Hd16DS)
4. Check how the product is passed to Google Analytics: 

        gtag('event', 'view_item', {
            'items': [{
                'id': 'bundle-one',
                'name': 'Product Bundle',
                'category': 'Accessories',
                'price': '0',
            }]
        });

### Expected behavior:

The cost of the bundle should reflect the items added to the bundle

### Actual behavior:

The cost of the bundle is universally defined as 0 despite the bundled items 

### Additional details:

Originally reported in 5366001-zen

**Comment by nicholasjeberlin** [comment]
39324603-hc

**Comment by tomalec** [comment]
This extension simply looks up the [`$product->get_price()`](https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/fd97569b48b17a126541bb62096b74a3f61ccafa/includes/class-wc-google-gtag-js.php#L424).

In the case of individually priced bundles, the Product Bundles extension sets its price to `0`. 
I'm not sure why. To me, it makes the rest of the ecosystem consider the bundle's price as such. :shrug: 
Whereas the cart uses a different value.

So we can make a specific integration between those two extensions and use `$product->get_bundle_price()` if available.
But to me, it's not a very sustainable solution. The "product price" is something exposed to everybody in Woo, and other extensions can contribute there. Whereas "bundle price" is something specific only to that extension.

@woocommerce/somewherewarm Can you suggest the desired way to fetch the effective price of a bundle as a product?

**Comment by jimjasson** [comment]
Hey @tomalec, 

Thank you for the ping! 

The `get_price` function returns the Regular/Sale price of a product. For Bundles without a Regular/Sale price, this is `0` -- the same happens for a Simple Product without a Regular/Sale price.

In contrast to Simple and Variable Products, Bundles do not always have a pre-defined price -- the price of the Bundle depends on the configuration the user makes. Therefore, it is not possible to define a function that reliably returns the Bundle's price, as this price can be dynamic. That's why, we created some helper functions which can return the price of the cheapest and most expensive configuration. 

I think that in your case, it makes sense to retrieve the price of the cheapest configuration via: `$product->get_bundle_price( 'min')`

Keep in mind that other product types have unique price functions as well. For example, the core Variable Products have the `get_variation_prices` function that is often used to determine what is the cheapest and most expensive price that a Variable Product can be bought at.

**Comment by tomalec** [comment]
Thank you for the quick reply @jimjasson ! :) 

> ... Bundles do not always have a pre-defined price --  the price of the Bundle depends on the configuration the user makes. Therefore, it is not possible to define a function that reliably returns the Bundle's price, as this price can be dynamic. That's why, we created some helper functions which can return the price of the cheapest and most expensive configuration.

If the bundle does not have a price, or it's unknown or non-numeric, I'd expect `get_price` to return `null`, not `0`. I'd read price == 0 as free, which in most cases, is not true.

How does that behave during the checkout? Then the bundle needs to have a definite price to know how much to charge. 

> Keep in mind that other product types have unique price functions as well. For example, the core Variable Products 

For variable products, it's `get_price` seems to return the minimum price. Which to me, looks like a reasonable fallback for consumers not expecting a range. What is the reason not to follow that behavior in Bundles?

I'd like to avoid implementing specific GoogleAnalyticsIntegration-to-ProductBundles integration if possible. In general, with the combinatorial explosion on top of the number of extensions in the ecosystem, such an approach does not scale well.
I imagine there could be yet another extension in our ecosystem that introduces a range of prices for a product. So, I'd rather have a consistent approach for the entire class of such extensions.
For example:
1. Return the minimum of the range from `get_price`, or
2. return `null` and add `get_price_range` method - regardless of your extension name. Variable products could implement it as well.

**Comment by jimjasson** [comment]
> How does that behave during the checkout? Then the bundle needs to have a definite price to know how much to charge.

On the checkout page, customers have already selected which items they'd like to include in the Bundle and therefore the price is calculated based on their selections.

Think of the following example: 
- A Bundle contains a mandatory item, "Shirt" which costs $30, and an optional item, "Extra buttons" which costs $5.
- If a customer selects only the mandatory item, the total price will be $30. If the optional item is selected as well, then the total price is $35. If the customer selects the optional item twice, the total price would be $40.

It is not possible to predict the price of this Bundle when it is displayed in the Catalog or the search results, as it depends on customers' configuration. The best solution here would be to display a "From $30" price, which is the price of the minimum configuration. 

> For variable products, it's get_price seems to return the minimum price. Which to me, looks like a reasonable fallback for consumers not expecting a range. What is the reason not to follow that behavior in Bundles?

The difference is that the parent Bundle item can also have its own Regular/Sale price, while the Variable Product always depends on the price of its variations. 

For example: 
- A Bundle can be a laptop and the bundled items could be optional accessories. In this case, the Bundle would have a Regular Price which would be equal to the laptop's price, and `get_price` should be able to get you this price. 
- A pick-and-mix Bundle where customers could select any 4 Hoodies. In this case, the Bundle itself is not a separate product, and therefore it doesn't have a Regular price. Its price depends only on which Hoodies customers select. 

P.S.: By Regular Price, I am referring to the price that is set under Product Data > General. 

> I'd like to avoid implementing specific GoogleAnalyticsIntegration-to-ProductBundles integration if possible. In general, with the combinatorial explosion on top of the number of extensions in the ecosystem, such an approach does not scale well.
I imagine there could be yet another extension in our ecosystem that introduces a range of prices for a product. So, I'd rather have a consistent approach for the entire class of such extensions.

I hear you! If you have a look at the integrations we maintain in Product Bundles, for example, you will notice that we have created a separate integration file for each plugin. 

Generally, you are right that we should aim for the least code to ensure compatibility between plugins. However, even if WooCommerce or some plugins introduced a specific function, as the `get_price_range` you suggest, there could still be other plugins that continue to work in a different way. In an open-source environment, one cannot expect that everyone's code follows the same conventions, as different plugins are written by different teams that do not always collaborate with each other. Even if we introduced a `get_price_range` for the purpose of this integration, how would every team, that wrote a plugin that introduced a new product type, know that this is the function they need to use? :) 

In any case, if you feel that such a function would be useful to the ecosystem, you are more than welcome to submit a PR with your idea to the WooCommerce core repo and continue this discussion there for visibility.

As for `get_price`, I do not agree that it should return `null` when the Bundle doesn't have a Regular/Sale price or fall back to the minimum price. As I explained above, there are many cases in which the Bundle should or shouldn't have a Regular price and these functions should return something consistent; `0` when a Regular/Sale is not set and the defined Regular Price otherwise.

Since this eventually boils down to each specific case, our approach would be to somehow allow merchants to decide which Structured Data format to use depending on each case. Here is what we are thinking: ![BGmyD3.png](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/25955483/ce074afb-6910-425b-85bc-c0d0c51e79df)

**Comment by tomalec** [comment]
Thank you for such elaborate explanations and examples that help me a lot :pray:  :) 

> On the checkout page, customers have already selected which items they'd like to include in the Bundle and therefore the price is calculated based on their selections.

And what method is used to fetch the bundle's price there? (sorry, I know I can look at the source code myself. But I'd like to avoid my human brain parser misinterpreting something ;) )

Speaking of examples, 
1. I'm trying to recreate the case from the OP.
   - Created a new product bundle ("Sunny days"), with regular & sale price empty
   - Added two bundled products using basic settings:
      - Sunglasses min=max=default quantity=1, priced individually, discount 5%
      - Cap min=max=default quantity=1, priced individually, discount 7%
   - Regular prices of the sunglasses and the cap are 90 and 18
2. Shirt + buttons
3. Laptop + accessories
4. Pick&mix 4 hoodies

I'll present my mental model and vocabulary, as given your explanations, it probably does not match the implementation and is the source of my confusion :) 

- "Product Bundle" 
   - :information_desk_person: Packaged offering to customers. 
   - :technologist: a WooCommerce product of type `bundle` + set (empty or not) of its _bundled products_
- "Bundle base" 
   - :information_desk_person: The package or the base of that offering. 
   - :technologist: a WooCommerce product of type `bundle`
- "Bundled Product" 
   - :information_desk_person: Individual parts of the offering. 
   - :technologist: a WooCommerce product added to the product of type `bundle` under **Product data > Bundled Products**
- "Variable product"
   - :information_desk_person: An (abstract) group of products sharing common characteristics, distinct on some attributes
   - :technologist: a WooCommerce product of type `variable`
- "Variation product"
   - :information_desk_person: A n individual real product, and element of `variable` group with specific set of attributes
   - :technologist: a variation of attributes added at **Product data > Variations** of a _Variable product_
- "Price" 
   - :information_desk_person: the amount of money a customer will be charged when they add a specific thing to their cart and complete the checkout process
- "Price range" 
   - :information_desk_person: A range of _price_s for a group of buyable things.

The _bundle base_ may have its own _price_ or not. _Product bundle_ may contain an empty, defined, or variable set of _bundled products_. _Bundled products_ may contribute/add their individual _prices_ to the total _product bundle_ _price_, or not. 

An instance of a _product bundle_ can be added to the cart, and its _price_ is dependent on the actual composition of _products_ selected by a customer and the configuration of those provided by the merchant.

The _variable_ product instance cannot be added to the cart as it's just a group of actual products. Each _variation_ has a specific single _price_

"_Bundle_", "_base_", and "_bundled_" are the concepts introduced by the Product Bundles extension. All the rest are generic to WooCommerce Core and ecosystem.

With the above model, the way I understood prices is:
- The "_Variable_" product does not have a _price_ as technically, you cannot buy an instance of _variable product_, you can only buy an instance of _variation_. With the sample data set, you cannot buy just a "Hoodie" - you can only buy "color, logo, Hoodie". So, a _variable product_'  can have a _price range_ of its _variations_. Personally, I'd find `get_price` equal to `null` or the minimum of that range a reasonable solution.
- Every instance of simple product or _variation_ has a defined _price_, where `0` means free of charge.
- Every instance of _product bundle_ has a static single _price_ = price of its _bundle base_ + _price_s of all its _bundled products_ selected.
- _Product bundle_ can be 
   - a single buyable thing with a static _price_, if it has an empty or static set of _bundled product_s . You can buy "Sunny days".
   - a group of buyable things with a static _price_, if it has no or a static set of individually priced _bundled product_s . You can buy "Pick&mix 4 hoodies for 160$".
   - a group of buyable things with different prices. "Laptop + accessories", "shirt + add-ons", "Pick&mix 4 hoodies for 25%off"
  So _Product bundle_  can have a _price range_ of its combinations: `[base price + cheapest combination, base price + most expensive combination]`

`$product->get_price()` means to me: "get the price of the product".  If the `$product` happens to be a _product bundle_. It means get the price of the product bundle = price of the bundle base + price of individually priced bundled products. 

Then, applying it to our examples, 
1. "Sunny days"
   - bundle base price = 0
   - bundled product prices = sunglasses `90*0.95`, cap `18*0.93`
   - product bundle price = 100.38
2. "Shirt + addons"
   - bundle base price = 30
   - bundled product prices = fancy button (0 to 10) 5, custom embroidery 15
   - product bundle price = [30, 95]
3. "Laptop + accessories"
   - bundle base price = 1000
   - bundle product prices = RAM (0 to 2) 100, extra SSD (0 to 1) 300, any peripherals 5%off
   - product bundle price = [1000, ∞]
     
 
 > As for `get_price`, I do not agree that it should return `null` when the Bundle doesn't have a Regular/Sale price or fall back to the minimum price. As I explained above, there are many cases in which the Bundle should or shouldn't have a Regular price and these functions should return something consistent; `0` when a Regular/Sale is not set and the defined Regular Price otherwise.
 
 Thanks to your examples. I now understand that there are cases when the Regular/Sale _base price_ should be empty or `0` and there are cases where it should be greater than `0`.
 
And I think my entire confusion comes from the fact that in the actually Product Bundles implementation 
- `$product->get_price()` does not return the product's price. It returns the **_bundle base price_** 
- `$product->get_bundle_price()` returns the product's price. 
 
Is that correct?

**Comment by tomalec** [comment]
>  there could still be other plugins that continue to work in a different way. In an open-source environment, one cannot expect that everyone's code follows the same conventions, as different plugins are written by different teams that do not always collaborate with each other. Even if we introduced a `get_price_range` for the purpose of this integration, how would every team, that wrote a plugin that introduced a new product type, know that this is the function they need to use? :)

Sure we cannot fix the other participants of the ecosystem. But we can change the way we participate to 1. make our life easier, 2. pave the path, and act as an example for the community. Plus we can improve the platform - make a proposal to WooCommerce Core, as you mentioned :godmode: :) 

If we sometimes fall into the pitfalls of cross-extension integration, 3PDs will struggle more, as they don't have such ease to peek into other extension sources and collaborate together.
So we cannot expect that everyone's code will follow the Product Bundle's conventions. All we can expect is that they will follow WooCommerce Core convention.

The way I understand it (possibly wrong, correct me please) is that the WooCommerce convention is that `$product->get_price` is the price of the thing (stored in `$product`) that a customer is about to purchase.
Whereas if my conclusions in the comment above (https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/219#issuecomment-1785999065) are correct, we're storing there "a price of Foo", where Foo is our own concept of the base price, which is just a subset of the entire thingy the customer is about to purchase. An implementation detail that we cannot expect the community to know.

**Comment by tomalec** [comment]
> Since this eventually boils down to each specific case, our approach would be to somehow allow merchants to decide which Structured Data format to use depending on each case. Here is what we are thinking:

:) sounds cool. This means with that feature, merchants will be able to specify what they mean by product bundle price on their own. Then probably we in Googly Analytics integrations could pick that up from the structured data. :)

**Comment by tomalec** [comment]
Getting back to actually addressing the original issue

The above elaborate and idealistic imaginations would boil down to two possible solutions:
1. Implement `get_price_range` in WooCommerce Core, and shift Variable Products and Product Bundles to use it, then here use minimum price.
   From my very brief understanding, it seems to make sense. There are probably more extensions using that concept, and instead of hardcoding and limiting aggregated offers to variations only https://github.com/woocommerce/woocommerce/blob/d663c304c92794e114fa052d7d7ed0b22a4037e2/plugins/woocommerce/includes/class-wc-structured-data.php#L221-L243
   However, it would require long discussions and explorations with the broader community, quite an effort to implement in Core, and Product Bundles. Then we'll need to wait few months for our L-2 policy.
2. Redesigning the bundle base concept in WC PB, just to match my fantasy described above ;) `get_price = get_bundle_price` & `get_base_price = old_get_price`.
      First of all, I'm not sure if my mental model actually makes any sense to the entire WC PB extension. And does not have any serious drawbacks. Then It's probably a bit revolutionary change at this stage of the extension.
    But maybe worth considering. I'm definitely very interested in your comments here.

Then we have the more down-to-earth solution, to work around it here in WC GAI. I'm still not sure how.

3. Use `$product->get_bundle_price()` if the `$product` is and instance of `WC_Product_Bundle`. Or fetch the minimum price otherwise.
   Is that the correct way to fetch such a price?
   It would probably be very quick to implement, test, and ship. However, IMHO, it's not the finest courtesy to the ecosystem, to patch it here because we can and expect all the other combinations of extensions in the wild to follow. But also, it's not very serious and profound either. It's just explicit integration :) 

Then it's the quickies way:

4. We can also, say that it's not a bug. Per the WC Product Bundles model the product bundle price on the product page is not fully determined, therefore `get_price() == 0` is correct and expected behavior. As such should be sent to Google Analytics. Close as not a bug, won't fix.
   Even though I like the simplest solutions, I'm not in favor of this one. 
   The "Sunny days" bundle price range is between 100.38 and 100.38, it can be both for no more no less than 100.38. The product page visually displays that price. Therefore I see no reason, why we should report to Google Analytics, that a customer was previewing a product page of a product worth `0`.

**Comment by jimjasson** [comment]
Hey @tomalec, 

Thank you for your elaborate exploration! 

I think that it is clear by now that what is Product Bundles strength is also its challenge; it offers way too many configuration possibilities to merchants, to the point where deterministically defining a proper schema is not entirely possible. 

So far, we have treated the "base" Bundle product as a Simple Product -- therefore, if you set bundled items aside for a moment, all the pricing functions that are used on a Simple Product can be used on a "base" Bundle product and return the same results. 

Now, adding bundled items into the mix, a new question emerges; What is the price of this product as a whole?

The answer to this is: "We don't know". We could never answer this reliably, as depending on which bundled items are selected and at which quantity, the product price differs. The only question, _we can_ answer is: "What can this product potentially cost?". This is why we created the Bundle-specific pricing functions that return the price of the cheapest and most expensive configuration. 

I agree with you, that ideally, all configurable product types would share the same function to determine what is the price range of the product. But, as you correctly mentioned, such a big change in the way the WooCommerce abstract product is defined would require a lot of discussion until it is implemented. 

The way I see it, the best way to integrate the two plugins is to create a specific integration file and implement this:

> Use $product->get_bundle_price() if the $product is and instance of WC_Product_Bundle. Or fetch the minimum price otherwise.

`$product->get_bundle_price( 'min')` is the optimal way to get the price of the cheapest Bundle configuration. Calling the same function with a `max` argument will get you the price of the most expensive configuration, given that it is not infinite.

Count me in if you have any specific technical questions or need help reviewing the final integration file!

> And what method is used to fetch the bundle's price there? (sorry, I know I can look at the source code myself. But I'd like to avoid my human brain parser misinterpreting something ;) )

When you add a Bundle to the cart, one cart item is created for the "base" Bundle product and one cart item is created for each bundled item. Then, it is up to WooCommerce core to calculate cart/order totals -- the calculation is the same as if these products were added to the cart individually. 

If you are referring to the aggregated total that shows up next to the "base" Bundle item in the cart/checkout, then this price is only for display and it is calculated in `WC_PB_Display::cart_item_subtotal` inside `woocommerce-product-bundles/includes/class-wc-pb-display.php`.
